/alive and /ready health checks do not consider subsystem health

### Discussed in https://github.com/eclipse-ditto/ditto/discussions/1861

<details>
<summary>Discussion details</summary>

<div type='discussions-op-text'>

<sup>Originally posted by **dimabarbul** January 12, 2024</sup>
We deploy Ditto (tried 3.4.0 and 3.4.4) using helm chart.  Sometimes, for example, during cluster upgrade, we face following issue.

Things service loses connection to mongo (this is expected as during cluster upgrade mongo might be temporarily unavailable). But later when mongo is back online things service does not reconnect. In logs we can see errors `com.mongodb.MongoTimeoutException: Timed out after 30000 ms while waiting for a server that matches ReadPreferenceServerSelector{readPreference=primary}. Client view of cluster state is {type=REPLICA_SET, srvResolutionException=com.mongodb.MongoConfigurationException: Failed looking up SRV record for '_mongodb._tcp.default-psmdb-db-rs0.databases.svc.cluster.local'.` But /alive and /ready health check endpoints return OK. Request to http://ditto-gateway:8080/health returns 503 as expected. We waited for more than an hour and the issue did not go away.

The only way that we found where we can see this error (apart from logs) is request to http://ditto-things:8080/status/health. The problem with this approach is that port 8080 is not declared in helm chart and thus is not accessible for k8s health probes.

Restarting the pod fixes the issue, but the problem is that this restart has to be manual as k8s does not see any issues.

<details>
<summary>Logs from things service that might be related</summary>
{"@timestamp":"2024-01-11T12:52:00.028+01:00","@version":"1","message":"Stream failed","logger_name":"org.eclipse.ditto.internal.utils.persistentactors.cleanup.PersistenceCleanupActor","thread_name":"ditto-cluster-pekko.actor.default-dispatcher-14","level":"ERROR","level_value":40000,"stack_trace":"java.util.concurrent.CompletionException: com.mongodb.MongoTimeoutException: Timed out after 30000 ms while waiting for a server that matches ReadPreferenceServerSelector{readPreference=primary}. Client view of cluster state is {type=REPLICA_SET, srvResolutionException=com.mongodb.MongoConfigurationException: Failed looking up SRV record for '_mongodb._tcp.default-psmdb-db-rs0.databases.svc.cluster.local'., servers=[]\n\tat java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332)\n\tat java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:347)\n\tat java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:636)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)\n\tat java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162)\n\tat scala.concurrent.impl.FutureConvertersImpl$CF.apply(FutureConvertersImpl.scala:26)\n\tat scala.concurrent.impl.FutureConvertersImpl$CF.apply(FutureConvertersImpl.scala:23)\n\tat scala.concurrent.impl.Promise$Transformation.run(Promise.scala:484)\n\tat scala.concurrent.ExecutionContext$parasitic$.execute(ExecutionContext.scala:222)\n\tat scala.concurrent.impl.Promise$Transformation.submitWithValue(Promise.scala:429)\n\tat scala.concurrent.impl.Promise$DefaultPromise.submitWithValue(Promise.scala:338)\n\tat scala.concurrent.impl.Promise$DefaultPromise.tryComplete0(Promise.scala:285)\n\tat scala.concurrent.impl.Promise$Transformation.run(Promise.scala:504)\n\tat scala.concurrent.ExecutionContext$parasitic$.execute(ExecutionContext.scala:222)\n\tat scala.concurrent.impl.Promise$Transformation.submitWithValue(Promise.scala:429)\n\tat scala.concurrent.impl.Promise$DefaultPromise.submitWithValue(Promise.scala:338)\n\tat scala.concurrent.impl.Promise$DefaultPromise.tryComplete0(Promise.scala:285)\n\tat scala.concurrent.impl.Promise$DefaultPromise.tryComplete(Promise.scala:278)\n\tat scala.concurrent.Promise.tryFailure(Promise.scala:117)\n\tat scala.concurrent.Promise.tryFailure$(Promise.scala:117)\n\tat scala.concurrent.impl.Promise$DefaultPromise.tryFailure(Promise.scala:104)\n\tat org.apache.pekko.stream.impl.SeqStage$$anon$3.onUpstreamFailure(Sinks.scala:274)\n\tat org.apache.pekko.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:537)\n\tat org.apache.pekko.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:400)\n\tat org.apache.pekko.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:662)\n\tat org.apache.pekko.stream.impl.fusing.ActorGraphInterpreter$SimpleBoundaryEvent.execute(ActorGraphInterpreter.scala:71)\n\tat org.apache.pekko.stream.impl.fusing.ActorGraphInterpreter$SimpleBoundaryEvent.execute$(ActorGraphInterpreter.scala:67)\n\tat org.apache.pekko.stream.impl.fusing.ActorGraphInterpreter$BatchingActorInputBoundary$OnError.execute(ActorGraphInterpreter.scala:92)\n\tat org.apache.pekko.stream.impl.fusing.GraphInterpreterShell.processEvent(ActorGraphInterpreter.scala:637)\n\tat org.apache.pekko.stream.impl.fusing.ActorGraphInterpreter.org$apache$pekko$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:813)\n\tat org.apache.pekko.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:831)\n\tat org.apache.pekko.actor.Actor.aroundReceive(Actor.scala:547)\n\tat org.apache.pekko.actor.Actor.aroundReceive$(Actor.scala:545)\n\tat org.apache.pekko.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:729)\n\tat org.apache.pekko.actor.ActorCell.receiveMessage(ActorCell.scala:590)\n\tat org.apache.pekko.actor.ActorCell.invoke(ActorCell.scala:557)\n\tat org.apache.pekko.dispatch.Mailbox.processMailbox(Mailbox.scala:280)\n\tat org.apache.pekko.dispatch.Mailbox.run(Mailbox.scala:241)\n\tat kamon.instrumentation.executor.ExecutorInstrumentation$InstrumentedForkJoinPool$TimingRunnable.run(ExecutorInstrumentation.scala:698)\n\tat org.apache.pekko.dispatch.ForkJoinExecutorConfigurator$PekkoForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:57)\n\tat java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)\n\tat java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)\n\tat java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)\n\tat java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)\n\tat java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)\nCaused by: com.mongodb.MongoTimeoutException: Timed out after 30000 ms while waiting for a server that matches ReadPreferenceServerSelector{readPreference=primary}. Client view of cluster state is {type=REPLICA_SET, srvResolutionException=com.mongodb.MongoConfigurationException: Failed looking up SRV record for '_mongodb._tcp.default-psmdb-db-rs0.databases.svc.cluster.local'., servers=[]\n\tat com.mongodb.internal.connection.BaseCluster.createTimeoutException(BaseCluster.java:430)\n\tat com.mongodb.internal.connection.BaseCluster.handleServerSelectionRequest(BaseCluster.java:325)\n\tat com.mongodb.internal.connection.BaseCluster.access$700(BaseCluster.java:69)\n\tat com.mongodb.internal.connection.BaseCluster$WaitQueueHandler.run(BaseCluster.java:505)\n\tat java.base/java.lang.Thread.run(Thread.java:840)\n","sourceThread":"Thread-37","pekkoAddress":"pekko://ditto-cluster@192.168.40.197:2551","pekkoSource":"pekko://ditto-cluster/user/thingsRoot/persistenceCleanup"}
{"@timestamp":"2024-01-11T12:52:00.029+01:00","@version":"1","message":"Stream failed or shutdown. Next stream in <PT6M26.71875S> starting from <>","logger_name":"org.eclipse.ditto.internal.utils.persistentactors.cleanup.PersistenceCleanupActor","thread_name":"ditto-cluster-pekko.actor.default-dispatcher-14","level":"INFO","level_value":20000,"sourceThread":"ditto-cluster-pekko.actor.default-dispatcher-11","pekkoAddress":"pekko://ditto-cluster@192.168.40.197:2551","pekkoSource":"pekko://ditto-cluster/user/thingsRoot/persistenceCleanup"}
{"@timestamp":"2024-01-11T12:52:07.879+01:00","@version":"1","message":"Sending the health of this system as requested: StatusInfo [status=DOWN, details=[StatusDetailMessage [level=ERROR, message=\"See detailed messages for: persistence.\"]], children=[StatusInfo [status=DOWN, details=[StatusDetailMessage [level=ERROR, message=\"com.mongodb.MongoTimeoutException: Timed out after 30000 ms while waiting for a server that matches WritableServerSelector. Client view of cluster state is {type=REPLICA_SET, srvResolutionException=com.mongodb.MongoConfigurationException: Failed looking up SRV record for '_mongodb._tcp.default-psmdb-db-rs0.databases.svc.cluster.local'., servers=[]\"]], children=[], label=persistence]], label=null]","logger_name":"org.eclipse.ditto.internal.utils.health.status.StatusSupplierActor","thread_name":"ditto-cluster-pekko.actor.default-dispatcher-6","level":"INFO","level_value":20000,"sourceThread":"Thread-38","pekkoAddress":"pekko://ditto-cluster@192.168.40.197:2551","pekkoSource":"pekko://ditto-cluster/user/statusSupplier"}
{"@timestamp":"2024-01-11T12:52:12.009+01:00","@version":"1","message":"Timed out after 30000 ms while waiting for a server that matches WritableServerSelector. Client view of cluster state is {type=REPLICA_SET, srvResolutionException=com.mongodb.MongoConfigurationException: Failed looking up SRV record for '_mongodb._tcp.default-psmdb-db-rs0.databases.svc.cluster.local'., servers=[]","logger_name":"org.eclipse.ditto.internal.utils.persistence.mongo.MongoHealthChecker","thread_name":"ditto-cluster-pekko.actor.default-dispatcher-11","level":"ERROR","level_value":40000,"stack_trace":"com.mongodb.MongoTimeoutException: Timed out after 30000 ms while waiting for a server that matches WritableServerSelector. Client view of cluster state is {type=REPLICA_SET, srvResolutionException=com.mongodb.MongoConfigurationException: Failed looking up SRV record for '_mongodb._tcp.default-psmdb-db-rs0.databases.svc.cluster.local'., servers=[]\n\tat com.mongodb.internal.connection.BaseCluster.createTimeoutException(BaseCluster.java:430)\n\tat com.mongodb.internal.connection.BaseCluster.handleServerSelectionRequest(BaseCluster.java:325)\n\tat com.mongodb.internal.connection.BaseCluster.access$700(BaseCluster.java:69)\n\tat com.mongodb.internal.connection.BaseCluster$WaitQueueHandler.run(BaseCluster.java:505)\n\tat java.base/java.lang.Thread.run(Thread.java:840)\n","sourceThread":"Thread-39","pekkoAddress":"pekko://ditto-cluster@192.168.40.197:2551","pekkoSource":"pekko://ditto-cluster/user/thingsRoot/healthCheckingActor/persistence"}
</details></div>

</details>

As discussed, it would be good to create class that would check subsystem health and could be used as Pekko management health check.

@thjaeckle I'm not sure about if the class should be included by default to liveness (or readiness) checks or this possibility should be just documented. Could you, please, advise?

--- a/connectivity/service/src/main/resources/connectivity.conf
+++ b/connectivity/service/src/main/resources/connectivity.conf
@@ -1137,6 +1137,9 @@ pekko {
       "pekko-contrib-mongodb-persistence-connection-snapshots"
     ]
   }
+  management.health-checks.liveness-checks {
+    subsystem-health = "org.eclipse.ditto.internal.utils.health.SubsystemHealthCheck"
+  }
 }
 
 include "ditto-protocol-subscriber.conf"

--- a/gateway/service/src/main/resources/gateway.conf
+++ b/gateway/service/src/main/resources/gateway.conf
@@ -575,6 +575,10 @@ pekko {
   management.health-checks.readiness-checks {
     gateway-http-readiness = "org.eclipse.ditto.gateway.service.health.GatewayHttpReadinessCheck"
   }
+
+  management.health-checks.liveness-checks {
+    subsystem-health = "org.eclipse.ditto.internal.utils.health.SubsystemHealthCheck"
+  }
 }
 
 include "ditto-protocol-subscriber.conf"

--- a/internal/utils/health/pom.xml
+++ b/internal/utils/health/pom.xml
@@ -70,6 +70,11 @@
             <artifactId>slf4j-simple</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.awaitility</groupId>
+            <artifactId>awaitility</artifactId>
+            <scope>test</scope>
+        </dependency>
     </dependencies>
 
 </project>

--- a/internal/utils/health/src/main/java/org/eclipse/ditto/internal/utils/health/SubsystemHealthCheck.java
+++ b/internal/utils/health/src/main/java/org/eclipse/ditto/internal/utils/health/SubsystemHealthCheck.java
@@ -1,0 +1,75 @@
+/*
+ * Copyright (c) 2024 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.ditto.internal.utils.health;
+
+import java.time.Duration;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Supplier;
+
+import org.apache.pekko.actor.ActorSelection;
+import org.apache.pekko.actor.ActorSystem;
+import org.apache.pekko.pattern.Patterns;
+
+/**
+ * Health check supplier for Pekko Management checking whether the root actor's health checking actor reports success.
+ * The health check will report failure if health checking actor replies with status DOWN or times out.
+ */
+public final class SubsystemHealthCheck implements Supplier<CompletionStage<Boolean>> {
+
+    /**
+     * The message to ask the health checking actor.
+     */
+    private static final RetrieveHealth RETRIEVE_HEALTH_ASK_MESSAGE = RetrieveHealth.newInstance();
+
+    /**
+     * Default timeout for waiting for response from health checking actor.
+     */
+    private static final Duration TIMEOUT = Duration.ofSeconds(30);
+
+    /**
+     * Health checking actor selection path.
+     */
+    private static final String HEALTH_CHECKING_ACTOR_PATH =
+            "/user/*Root/" + DefaultHealthCheckingActorFactory.ACTOR_NAME;
+
+    private final ActorSelection healthCheckingActor;
+    private final Duration timeout;
+
+    /**
+     * Constructs subsystem health check with default timeout.
+     *
+     * @param system actor system to check health of
+     */
+    public SubsystemHealthCheck(final ActorSystem system) {
+        this(system, TIMEOUT);
+    }
+
+    /**
+     * Constructs subsystem health check with custom timeout.
+     *
+     * @param system actor system to check health of
+     * @param timeout timeout
+     */
+    public SubsystemHealthCheck(final ActorSystem system, final Duration timeout) {
+        healthCheckingActor = system.actorSelection(HEALTH_CHECKING_ACTOR_PATH);
+        this.timeout = timeout;
+    }
+
+    @Override
+    public CompletionStage<Boolean> get() {
+        return Patterns.ask(healthCheckingActor, RETRIEVE_HEALTH_ASK_MESSAGE, timeout)
+                .handle((answer, throwable) -> answer instanceof StatusInfo statusInfo
+                        && statusInfo.getStatus() != StatusInfo.Status.DOWN);
+    }
+
+}

--- a/internal/utils/health/src/test/java/org/eclipse/ditto/internal/utils/health/SubsystemHealthCheckTest.java
+++ b/internal/utils/health/src/test/java/org/eclipse/ditto/internal/utils/health/SubsystemHealthCheckTest.java
@@ -1,0 +1,159 @@
+/*
+ * Copyright (c) 2024 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.ditto.internal.utils.health;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.time.Duration;
+
+import org.apache.pekko.actor.AbstractActor;
+import org.apache.pekko.actor.ActorRef;
+import org.apache.pekko.actor.ActorSystem;
+import org.apache.pekko.actor.Props;
+import org.apache.pekko.japi.pf.ReceiveBuilder;
+import org.apache.pekko.testkit.javadsl.TestKit;
+import org.awaitility.Awaitility;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class SubsystemHealthCheckTest {
+
+    private static final String TEST_ACTOR_READY = "ready";
+    private static final Duration HEALTH_CHECK_TIMEOUT = Duration.ofMillis(500);
+
+    /**
+     * Name of root actor in tests. Must end with "Root" to be recognized by {@link SubsystemHealthCheck}.
+     */
+    private static final String ROOT_ACTOR_NAME = "testRoot";
+
+    private ActorSystem actorSystem;
+    private SubsystemHealthCheck subsystemHealthCheck;
+
+    @Before
+    public void setUp() {
+        actorSystem = ActorSystem.create();
+        subsystemHealthCheck = new SubsystemHealthCheck(actorSystem, HEALTH_CHECK_TIMEOUT);
+    }
+
+    @After
+    public void tearDown() {
+        if (actorSystem != null) {
+            actorSystem.terminate();
+        }
+    }
+
+    @Test
+    public void testCheckReturnsTrueWhenHealthCheckingActorReturnsUp() {
+        new TestKit(actorSystem) {{
+            final var statusInfo = StatusInfo.fromStatus(StatusInfo.Status.UP);
+            actorSystem.actorOf(TestRootActor.props(statusInfo, getRef()), ROOT_ACTOR_NAME);
+
+            expectMsg(TEST_ACTOR_READY);
+
+            assertHealthCheckResult(subsystemHealthCheck, true);
+        }};
+    }
+
+    @Test
+    public void testCheckReturnsTrueWhenHealthCheckingActorReturnsUnknown() {
+        new TestKit(actorSystem) {{
+            final var statusInfo = StatusInfo.fromStatus(StatusInfo.Status.UNKNOWN);
+            actorSystem.actorOf(TestRootActor.props(statusInfo, getRef()), ROOT_ACTOR_NAME);
+
+            expectMsg(TEST_ACTOR_READY);
+
+            assertHealthCheckResult(subsystemHealthCheck, true);
+        }};
+    }
+
+    @Test
+    public void testCheckReturnsFalseWhenHealthCheckingActorReturnsDown() {
+        new TestKit(actorSystem) {{
+            final var statusInfo = StatusInfo.fromStatus(StatusInfo.Status.DOWN);
+            actorSystem.actorOf(TestRootActor.props(statusInfo, getRef()), ROOT_ACTOR_NAME);
+
+            expectMsg(TEST_ACTOR_READY);
+
+            assertHealthCheckResult(subsystemHealthCheck, false);
+        }};
+    }
+
+    @Test
+    public void testCheckReturnsFalseWhenHealthCheckingActorDoesNotExist() {
+        new TestKit(actorSystem) {{
+            assertHealthCheckResult(subsystemHealthCheck, false);
+        }};
+    }
+
+    private static void assertHealthCheckResult(SubsystemHealthCheck subsystemHealthCheck, boolean expected) {
+        final var healthCheckResultFuture = subsystemHealthCheck.get();
+        Awaitility.await()
+                .atMost(Duration.ofNanos(2 * HEALTH_CHECK_TIMEOUT.toNanos()))
+                .until(() -> healthCheckResultFuture.toCompletableFuture().isDone());
+        assertThat(healthCheckResultFuture.toCompletableFuture())
+                .isCompletedWithValue(expected);
+    }
+
+    private static class TestRootActor extends AbstractActor {
+
+        private TestRootActor(final StatusInfo statusInfo, final ActorRef actorToNotify) {
+            getContext().actorOf(TestHealthCheckingActor.props(statusInfo, actorToNotify),
+                    DefaultHealthCheckingActorFactory.ACTOR_NAME);
+        }
+
+        public static Props props(final StatusInfo statusInfo, final ActorRef actorToNotify) {
+            return Props.create(TestRootActor.class, statusInfo, actorToNotify);
+        }
+
+        @Override
+        public Receive createReceive() {
+            return ReceiveBuilder.create()
+                    .matchAny(this::unhandled)
+                    .build();
+        }
+
+    }
+
+    private static class TestHealthCheckingActor extends AbstractActor {
+
+        private final StatusInfo statusInfo;
+        private final ActorRef actorToNotify;
+
+        private TestHealthCheckingActor(final StatusInfo statusInfo, final ActorRef actorToNotify) {
+            this.statusInfo = statusInfo;
+            this.actorToNotify = actorToNotify;
+        }
+
+        public static Props props(final StatusInfo statusInfo, final ActorRef actorToNotify) {
+            return Props.create(TestHealthCheckingActor.class, statusInfo, actorToNotify);
+        }
+
+        @Override
+        public Receive createReceive() {
+            return ReceiveBuilder.create()
+                    .match(RetrieveHealth.class, this::returnStatusInfo)
+                    .build();
+        }
+
+        @Override
+        public void preStart() {
+            actorToNotify.tell(TEST_ACTOR_READY, getSelf());
+        }
+
+        private void returnStatusInfo(final RetrieveHealth command) {
+            getSender().tell(this.statusInfo, getSelf());
+        }
+    }
+
+}

--- a/policies/service/src/main/resources/policies.conf
+++ b/policies/service/src/main/resources/policies.conf
@@ -231,6 +231,10 @@ pekko {
       "pekko-contrib-mongodb-persistence-policies-snapshots"
     ]
   }
+
+  management.health-checks.liveness-checks {
+    subsystem-health = "org.eclipse.ditto.internal.utils.health.SubsystemHealthCheck"
+  }
 }
 
 pekko-contrib-mongodb-persistence-policies-journal {

--- a/things/service/src/main/resources/things.conf
+++ b/things/service/src/main/resources/things.conf
@@ -273,6 +273,10 @@ pekko {
     ]
   }
 
+  management.health-checks.liveness-checks {
+    subsystem-health = "org.eclipse.ditto.internal.utils.health.SubsystemHealthCheck"
+  }
+
 }
 
 pekko-contrib-mongodb-persistence-things-journal {

--- a/thingsearch/service/src/main/resources/search.conf
+++ b/thingsearch/service/src/main/resources/search.conf
@@ -340,6 +340,10 @@ pekko {
       "blocked-namespaces-aware",
     ]
   }
+
+  management.health-checks.liveness-checks {
+    subsystem-health = "org.eclipse.ditto.internal.utils.health.SubsystemHealthCheck"
+  }
 }
 
 search-dispatcher {

